### ISSUE TITLE
Specific mapping from JSON to NSObject property

### ISSUE THREAD

**Issue by vittoriom** [uacaps/NSObject-ObjectMap#9]
Automatic mapping is cool, but sometimes you just don't want to use the same names from the JSON for your object properties. 
Could you be interested in this feature? If yes, I will take care of implementing it in my spare time. I was thinking of something like creating a protocol and asking the object for a mapping between JSON fields and its properties. If you don't implement the protocol, everything will work as now, otherwise you can customize things a little bit more.
I was also thinking on creating another protocol for parsing and trasforming (or normalize) the values from JSON (or XML) before setting the properties, but this could be an issue on its own. 
Let me know what you think!
Thanks

**Comment by bennyguitar** [comment]
What was your plan for accomplishing this? I think it'd be cool if you basically had something like this:

``` objc
SingleObject *newObj = [NSObject objectOfClass:@"SingleObject" fromJSONData:jsonData withMap:@{@"_name":@"Name"}];
```

Or something similar. One thing I don't necessarily like about our current implementation is the fact that top-level arrays in JSON necessitate doing something in a custom init method. I believe that there should be no other methods or extra custom init stuff to necessitate this.

Of course this means that fortuitous coders would make a <code>+(NSDictionary)propertyMap</code> function or something similar, but I'd like not to have to rely on that.

**Comment by MatthewYork** [comment]
I am on board with this as well. One way to do things might be to implement a tagging or annotation-like system using C macros. This, to me, is the right way of going about it as it appears to be the least intrusive, and would not break previous implementations.

Something like this:
https://github.com/mattiaslevin/ObjectMapper

The benefit we would have is that we also map XML/SOAP instead of just JSON.

**Comment by vittoriom** [comment]
I have to study the library and its philosophy in the next days, I just found the project and I liked the idea of creating objects from JSON and XML automatically (but this could be expanded to other data formats as well, using some kind of abstract interface that each format would extend).
One thing I don't like in the current implementation is the class method added to NSObject for initializing instances.
I would rather do something like 

```
SimpleObject *sObj = [[SimpleObject alloc] initWithData:jsonData format:UAJSONFormat];
```

this would make things easier for the developer and more flexible too, in my opinion (for example, one could create the instance in different ways, and then set the properties in a second moment). The current implementation ties property initialization with instance creation and allocation, I don't like this very much.

As I said, I would create a protocol for mapping properties. I don't think it's a good idea to specify the mapping as an argument during initialization because it's unlikely that it would change in different calls for the same class. Classes (or developers) that choose to not implement the protocol will just get the default behavior (direct mapping from JSON/XML fields to properties).

As for what @MatthewYork suggests, I find annotations to be quite cool but I also think that macros should be avoided when possible, and in this case it's possible (my 2 cents).

**Comment by MatthewYork** [comment]
@vittoriom  Ah, so you are macro averse? ;-)

I am too, in most cases. If we can find a solution that is as clean (or almost as clean) but doesn't use them, I agree that would be ideal. 

Also, I just implemented those init methods and added unit tests for them. They seem to work pretty well from our tests. Check out

initWithJSONData:
initWithXMLData:

**Comment by vittoriom** [comment]
It's not that I hate macros, it's just that I prefer not to use them when there is an equivalent solution in code :)
I took a look at the project you posted, it looks like macros really make the developer's life easier in this case, but still what about those mapping with block macros? If this was even the only reason to not use macros for this purpose, it would be worth! 

My solution would be:
- Create a protocol that classes need to implement for mapping properties to JSON/XML fields.
- The protocol could be used for mapping array properties as well, with another method 

```
- (Class) classForArrayField:(NSString *)fieldName;
```

so that we don't have to specify array mappings in the weird init method anymore.
- For mapping / transforming with blocks (I find this pretty cool and I already hinted at this feature in my previous post), I still have to think about how to implement it, but a separate protocol would do the trick I guess.
